### PR TITLE
Issue 8236: Fixed the AppTracker class to use the ApplicationStateLis…

### LINKED ISSUE #8236: Deployed applications are not registered properly in mpHealth-2.0, which results in false positive checks
Sometimes, when applications are deployed into the running server via dropins, the mpHealth-2.0 feature does not register them properly, which results in a false positive health check response temporarily.

**Workaround**: Try accessing the Health Check endpoints again, or set the Kubernetes  readiness/liveness probe `initialDelaySeconds` field, so the endpoints are accessed, after the applications are fully registered.

```
[7/3/19 17:01:39:993 GMT] 00000043 id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       > addStartedApplication Entry  
                                                                                                               {com.ibm.wsspi.application.Application}={.installedByDropins=true, location=/WebSphere/Misc/jazz_build/jbe_flash182/jazz/buildsystem/buildengine/eclipse/build/dev/image/output/wlp/usr/servers/Health20TCKServer/dropins/tck.war, service.id=552, service.bundleid=72, service.scope=singleton, application.state=STARTED, service.factoryPid=com.ibm.ws.app.manager, name=tck, service.pid=com.ibm.ws.app.manager_23, type=war}
[7/3/19 17:01:39:993 GMT] 00000043 id=00000000 com.ibm.ws.microprofile.health.internal.AppTrackerImpl       3 addStartedApplication(): app added = tck
[7/3/19 17:01:39:993 GMT] 00000043 id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       < addStartedApplication Exit 
[7/3/19 17:01:39:994 GMT] 0000002a id=00000000 SessionContextRegistryImpl                                   I getSessionContext SESN0176I: A new session context will be created for application key default_host/tck
[7/3/19 17:01:39:995 GMT] 0000002a id=00000000 IDGeneratorImpl                                              I IDGeneratorImpl SESN0172I: The session manager is using the Java default SecureRandom implementation for session ID generation.
[7/3/19 17:01:40:050 GMT] 00000046 id=dea0e5c6 rofile.health20.internal.servlet.HealthCheckReadinessServlet > service Entry  
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletRequest31@ec4ba9a5
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletResponse31@36e4aee1
[7/3/19 17:01:40:050 GMT] 00000046 id=dea0e5c6 rofile.health20.internal.servlet.HealthCheckReadinessServlet > findHealthService Entry  
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletRequest31@ec4ba9a5
[7/3/19 17:01:40:050 GMT] 00000046 id=dea0e5c6 rofile.health20.internal.servlet.HealthCheckReadinessServlet < findHealthService Exit 
[7/3/19 17:01:40:050 GMT] 00000046 id=1b9f74da m.ws.microprofile.health20.internal.HealthCheck20ServiceImpl > performHealthCheck Entry  
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletRequest31@ec4ba9a5
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletResponse31@36e4aee1
                                                                                                               READY
[7/3/19 17:01:40:050 GMT] 00000046 id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       > getAppNames Entry 
[7/3/19 17:01:40:050 GMT] 00000046 id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       < getAppNames Exit  
                                                                                                               []
[7/3/19 17:01:40:050 GMT] 00000046 id=00000000 .microprofile.health.internal.HealthCheckHttpResponseBuilder > <init> Entry 
[7/3/19 17:01:40:050 GMT] 00000046 id=4c880154 .microprofile.health.internal.HealthCheckHttpResponseBuilder < <init> Exit  
                                                                                                               com.ibm.ws.microprofile.health20.internal.HealthCheck20HttpResponseBuilder@4c880154
[7/3/19 17:01:40:051 GMT] 00000046 id=00000000 roprofile.health20.internal.HealthCheck20HttpResponseBuilder > <init> Entry 
[7/3/19 17:01:40:051 GMT] 00000046 id=4c880154 roprofile.health20.internal.HealthCheck20HttpResponseBuilder < <init> Exit  
                                                                                                               com.ibm.ws.microprofile.health20.internal.HealthCheck20HttpResponseBuilder@4c880154
[7/3/19 17:01:40:051 GMT] 00000046 id=4c880154 roprofile.health20.internal.HealthCheck20HttpResponseBuilder > setHttpResponse Entry  
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletResponse31@36e4aee1
**[7/3/19 17:01:40:051 GMT] 00000046 id=4c880154 .microprofile.health.internal.HealthCheckHttpResponseBuilder > setJSONPayload Entry  
                                                                                                               {checks=[Ljava.lang.Object;@2df10f02, status=UP}**
                                                                                                               com.ibm.ws.webcontainer31.srt.SRTServletResponse31@36e4aee1
[7/3/19 17:01:40:051 GMT] 00000046 id=4c880154 .microprofile.health.internal.HealthCheckHttpResponseBuilder > getJSON Entry 
[7/3/19 17:01:40:051 GMT] 00000046 id=4c880154 .microprofile.health.internal.HealthCheckHttpResponseBuilder < getJSON Exit  
                                                                                                               com.ibm.ws.jsonsupport.internal.JSONJacksonImpl@4cd2e81
[7/3/19 17:01:40:052 GMT] 00000046 id=4c880154 .microprofile.health.internal.HealthCheckHttpResponseBuilder < setJSONPayload Exit 
[7/3/19 17:01:40:052 GMT] 00000046 id=4c880154 roprofile.health20.internal.HealthCheck20HttpResponseBuilder < setHttpResponse Exit 
[7/3/19 17:01:40:052 GMT] 00000046 id=1b9f74da m.ws.microprofile.health20.internal.HealthCheck20ServiceImpl < performHealthCheck Exit 
[7/3/19 17:01:40:052 GMT] 00000046 id=dea0e5c6 rofile.health20.internal.servlet.HealthCheckReadinessServlet < service Exit 
[7/3/19 17:01:40:057 GMT] 0000002a id=3e7ed5fa **com.ibm.ws.microprofile.health.internal.AppTrackerImpl       > onStartup Entry  
                                                                                                               [class org.eclipse.microprofile.health.tck.deployment.SuccessfulLiveness, class org.eclipse.microprofile.health.tck.deployment.FailedLiveness, class org.eclipse.microprofile.health.tck.deployment.SuccessfulReadiness, interface org.eclipse.microprofile.health.HealthCheck]**
                                                                                                               com.ibm.ws.webcontainer31.osgi.webapp.WebApp31@c4287805[tck#tck.war]
[7/3/19 17:01:40:058 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       > setAppModuleNames Entry  
                                                                                                               com.ibm.ws.webcontainer31.osgi.webapp.WebApp31@c4287805[tck#tck.war]
[7/3/19 17:01:40:058 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       > isOsgiApp Entry  
                                                                                                               com.ibm.ws.webcontainer31.osgi.webapp.WebApp31@c4287805[tck#tck.war]
[7/3/19 17:01:40:058 GMT] 0000002a id=00000000 com.ibm.ws.microprofile.health.internal.AppTrackerImpl       3 Servet context attr for key = osgi-bundlecontext, = null
[7/3/19 17:01:40:058 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       < isOsgiApp Exit  
                                                                                                               false
[7/3/19 17:01:40:058 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       > addAppModuleNames Entry  
                                                                                                               tck
                                                                                                               tck#tck.war
[7/3/19 17:01:40:058 GMT] 0000002a id=00000000 com.ibm.ws.microprofile.health.internal.AppTrackerImpl       3 addAppModuleNames(): modules added = {tck=[tck.war]} for app: tck
[7/3/19 17:01:40:058 GMT] 0000002a id=00000000 ws.microprofile.health.internal.AppTrackerImpl$AppModuleName > <init> Entry 
[7/3/19 17:01:40:058 GMT] 0000002a id=ddebbaef ws.microprofile.health.internal.AppTrackerImpl$AppModuleName < <init> Exit  
                                                                                                               com.ibm.ws.microprofile.health.internal.AppTrackerImpl$AppModuleName@ddebbaef
[7/3/19 17:01:40:058 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       < addAppModuleNames Exit  
                                                                                                               com.ibm.ws.microprofile.health.internal.AppTrackerImpl$AppModuleName@ddebbaef
[7/3/19 17:01:40:058 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       < setAppModuleNames Exit  
                                                                                                               com.ibm.ws.microprofile.health.internal.AppTrackerImpl$AppModuleName@ddebbaef
[7/3/19 17:01:40:059 GMT] 0000002a id=00000000 icroprofile.health.internal.AppTrackerServletContextListener > <init> Entry  
                                                                                                               com.ibm.ws.microprofile.health.internal.AppTrackerImpl$AppModuleName@ddebbaef
                                                                                                               com.ibm.ws.microprofile.health.internal.AppTrackerImpl@3e7ed5fa
[7/3/19 17:01:40:059 GMT] 0000002a id=bd2c3784 icroprofile.health.internal.AppTrackerServletContextListener < <init> Exit  
                                                                                                               com.ibm.ws.microprofile.health.internal.AppTrackerServletContextListener@bd2c3784
[7/3/19 17:01:40:059 GMT] 0000002a id=3e7ed5fa com.ibm.ws.microprofile.health.internal.AppTrackerImpl       < onStartup Exit 
[7/3/19 17:01:40:060 GMT] 0000002a id=bd2c3784 icroprofile.health.internal.AppTrackerServletContextListener > contextInitialized Entry  
                                                                                                               javax.servlet.ServletContextEvent[source=com.ibm.ws.webcontainer31.facade.ServletContextFacade31@e3b60a4a]
[7/3/19 17:01:40:060 GMT] 0000002a id=bd2c3784 icroprofile.health.internal.AppTrackerServletContextListener < contextInitialized Exit 
```

### PR DESCRIPTION
**Pull request by pgunapal**
…tener, instead of ServletContainerInitializer, to track the Application states.

The ServletContainerInitializer.startup() class does not work properly, due to timing issues in multi-threaded environments, such as OpenLiberty, where applications may be deployed and started at any given time. Re-designed the way how applications are registered and tracked for mpHealth, using the ApplicationStateListener interface.

Fixes #8236

### CODE REVIEW

**Comment by pgunapal** [comment]
#build

**Comment by pgunapal** [comment]
Fixes #8236

File: dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/internal/AppTrackerImpl.java [hunk 1]
```diff
@@ -118,50 +96,141 @@ public void onStartup(Set<Class<?>> arg0, ServletContext ctx) throws ServletExce
...
+            return;
+        }
 
-        WebAppConfig webAppConfig = isc.getWebAppConfig();
-        if (webAppConfig.isSystemApp()) {
-            Tr.debug(tc, "Detected system app so won't track for health check; appName = ", webAppConfig.getApplicationName());
-            return null;
+        try {
+            NonPersistentCache cache = appContainer.adapt(NonPersistentCache.class);
+            ApplicationClassesContainerInfo acci = (ApplicationClassesContainerInfo) cache.getFromCache(ApplicationClassesContainerInfo.class);
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/internal/AppTrackerImpl.java]: acci doesn't seem to be read anywhere else.

Comment (pgunapal) [reply dev/com.ibm.ws.microprofile.health/src/com/ibm/ws/microprofile/health/internal/AppTrackerImpl.java]: We need to check if the ApplicationClassesContainerInfo is in the container cache, before getting the application info from the container. I do a null check for acci in 118, and return if the application classes can't be found. If the application classes are found, than our Application Manager is in control of the application. For OSGI apps, (such as .eba or .esa applications, the acci would be null.

**Comment by pgunapal** [comment]
#build

**Comment by pgunapal** [comment]
#build

**Comment by pgunapal** [comment]
#build

**Comment by pgunapal** [comment]
#build

**Comment by pgunapal** [comment]
The build failures are related to invalid SOE platform string, all related mpHealth FATs and tests have passed.

**Pull request merged**